newton.go

```diff
@@ -0,0 +1,203 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package optimize
+
+import (
+	"errors"
+	"math"
+
+	"github.com/gonum/floats"
+	"github.com/gonum/matrix/mat64"
+)
+
+// Newton implements pure Newton's method without linesearch for unconstrained
+// minimization of functions that satisfy Gradient and Hessian interfaces.
+//
+// Newton's method iteratively forms a quadratic model to the objective
+// function f and tries to minimize this approximate model. It generates a
+// sequence of locations x_k by means of
+//  solve H_k d_k = -∇f_k for d_k,
+//  x_{k+1} = x_k + d_k,
+// where H_k is the Hessian matrix of f at x_k.
+//
+// Note that Newton's method is only locally convergent. This means that for
+// starting points close to a minimizer the sequence x_k converges rapidly with
+// a quadratic rate. However, away from a minimizer the quadratic approximation
+// may not be valid, the Hessian may not be positive definite and d_k may not
+// be a descent direction. In such case Newton fails with an error.
+//
+// For a globally convergent Hessian-based method, see ModifiedNewton. If the
+// Hessian matrix cannot be formed explicitly or if the computational cost of
+// its factorization is prohibitive, BFGS or L-BFGS methods can be used
+// instead.
+type Newton struct {
+	chol *mat64.TriDense
+	step *mat64.Dense
+}
+
+func (n *Newton) Init(loc *Location, f *FunctionInfo, xNext []float64) (EvaluationType, IterationType, error) {
+	dim := len(loc.X)
+	n.chol = resizeTriDense(n.chol, dim)
+	r, _ := n.step.Dims()
+	if dim < r {
+		n.step = n.step.View(0, 0, dim, 1).(*mat64.Dense)
+	} else if dim > r {
+		n.step.Grow(dim, 1)
+	}
+	return n.computeNextX(loc, xNext)
+}
+
+func (n *Newton) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
+	return n.computeNextX(loc, xNext)
+}
+
+func (n *Newton) computeNextX(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
+	pd := n.chol.Cholesky(loc.Hessian, false)
+	if !pd {
+		return NoEvaluation, NoIteration, errors.New("optimize: Hessian is not positive definite")
+	}
+	n.step.SolveCholesky(n.chol, mat64.NewDense(len(loc.Gradient), 1, loc.Gradient))
+	floats.AddScaledTo(xNext, loc.X, -1, n.step.RawMatrix().Data)
+	return FuncEvaluation | GradEvaluation | HessEvaluation, MajorIteration, nil
+}
+
+func (n *Newton) Needs() struct {
+	Gradient bool
+	Hessian  bool
+} {
+	return struct {
+		Gradient bool
+		Hessian  bool
+	}{true, true}
+}
+
+// ModifiedNewton implements a modified Newton's method for unconstrained
+// minimization of functions that satisfy Gradient and Hessian interfaces. It
+// is a globally convergent line-search method that can be applied from any
```

Comment:
We need to find an alternate expression to 'globally convergent'. That implies it will find a global minimum.

// Newton implements a modified Newton's method for Hessian-based 
// unconstrained minimization. It applies regularization when the 
// Hessian is non-positive-definite, and uses line-searches so it can
// converge to a local minimum from any starting point.